### DISCUSSION TITLE
Why is health check initializing a new instance of BackgroundWorker each time?

### DISCUSSION THREAD

**Discussion by Kampfmoehre** [dotnet/aspnetcore#54801] [Q&A]
Hi,

I have implemented `IHealthCheck` in a `BackgroundService` that is registered via `services.AddHostedService<MyWorker>()` and added to health checks via `services.AddHealthChecks().AddCheck<MyWorker>()`.

The health check is implemented in a way, that is indicating healthy once `ExecuteAsync` was called in the worker and until some event happened. However upon testing I saw that the worker was unhealthy despite running correctly. When debugging I could see, that the health check created a new instance of the worker every time.

Since I am using .NET 8 `AddHostedService` should register the worker as singleton. Why is the health check not recieving the singleton instance then, am I missing something?

I worked around that by changing the registration of the worker to the following:
```cs
services.AddSingleton<MyWorker>();
services.AddHostedService(sp => sp.GetRequiredService<MyWorker>());
```

With this the health checks works as expected, however I could not find any documentation or issues with this problem, so why is the "official" way not working as expected?

Here is the relevant code in case it is helpful. Note, the worker is defined in a NuGet package and then inherited in the actual project.
```cs
public abstract class MyWorkerBase : BackgroundService, IHealthCheck
{
  private bool _processStarted;

  public virtual Task<HealthCheckResult> CheckHealthAsync(
    HealthCheckContext context,
    CancellationToken cancellationToken = default)
  {
    if (!_processStarted)
    {
      return Task.FromResult(HealthCheckResult.Unhealthy("Process is not active."));
    }

    return Task.FromResult(HealthCheckResult.Healthy("Process is running."));
  }

  protected override async Task ExecuteAsync(CancellationToken stoppingToken)
  {
    // ...
    SomeEvent += SomeEventHandler;
    _processStarted = true;
    // ...
  }

  private async Task SomeEventHandler(object sender, EventArgs args)
  {
    _processStarted = false;
  }
```
And the main project
```cs
public class MyWorker : MyWorkerBase
{
  // ...
}

// ...
services.AddHostedService<MyWorker>();
services
  .AddHealthChecks()
  .AddCheck<MyWorker>("my-worker");
```

Second note: I am using a custom HealthCheckPublisher that writes the health check to a file since the project does not contain any HTTP server.
```cs
services.AddSingleton<IHealthCheckPublisher, HealthCheckPublisher>();
```

**Comment by martincostello** [comment]
Looking at the code, the `AddCheck<T>()` method uses `ActivatorUtilities.GetServiceOrCreateInstance<T>()` to resolve the service, so I'm not sure why the previously registered instance isn't being used:

https://github.com/dotnet/aspnetcore/blob/63c8031b6a6af5009b3c5bb4291fcc4c32b06b10/src/HealthChecks/HealthChecks/src/DependencyInjection/HealthChecksBuilderAddCheckExtensions.cs#L131

Does `MyWorker` have something about its declaration/implementation (attributes, multiple constructors) that causes `ActivatorUtilities` to behave differently?

**Reply by Kampfmoehre** [reply]
No, just a single constructor, no attributes. It is an AMQP consumer and the mentioned event handles consumer cancellation/shutdown.
The health check is implemented in an abstract base class, which also has a derived class with additional error handling in it and then another derived with the actual worker which implements the message handling. Only the latter is defined in the actual project but I don't think that should matter somehow.

The abstract base class receives another singleton instance (the AMQP connection) i nthe constrcutor and some options and a logger, nothing special.
